### ISSUE TITLE
low and high of torch.testing.make_tensor don't work correctly for bool and integral dtypes

### ISSUE THREAD

**Issue by pmeier** [pytorch/pytorch#96101]
TL;DR This issue proposes two changes to `torch.testing.make_tensor` for boolean and integral dtypes:

1. Don't ignore `low` and `high` for `torch.bool`
2. Correctly round floating point `low` and `high` for integral dtypes

---

For other dtypes the `low` and `high` are used to specify the upper and lower bound of the values generated by `torch.testing.make_tensor`:

https://github.com/pytorch/pytorch/blob/7a192cc51c7172860efd35413acf9a8b9aafd2c9/torch/testing/_creation.py#L150-L153

For `torch.bool` however, this is not the case:

https://github.com/pytorch/pytorch/blob/7a192cc51c7172860efd35413acf9a8b9aafd2c9/torch/testing/_creation.py#L141-L142

Intuitively, this makes sense since a boolean tensor can only support two values, and thus specifying anything else makes little sense. And that holds true until you are in a parametrized environment.

Imagine you want to test an operator like [`torch.atanh`](https://pytorch.org/docs/stable/generated/torch.atanh.html?highlight=atanh#torch.atanh). The only valid input values are `(-1, 1)`, which translates to a tensor full of `False` for `torch.bool` (or `0` for the integral dtypes). However,

```pycon
>>> eps = 1e-4
>>> torch.manual_seed(0)
>>> torch.testing.make_tensor(10, dtype=torch.bool, device="cpu", low=-(1 - eps), high=1 - eps)
tensor([False,  True, False, False,  True, False, False,  True,  True, False])
```

Ok, let's honor `low` and `high` for `torch.bool` as well. Does it work now?

```python
for dtype in [torch.float32, torch.int32, torch.uint8, torch.bool]:

    torch.manual_seed(0)
    input = torch.testing.make_tensor(10, dtype=dtype, device="cpu", low=-(1 - eps), high=1 - eps)

    output = torch.atanh(input)

    if (torch.isinf(output) | torch.isnan(output)).any():
        print(f"FAIL: {dtype}")
    else:
        print(f"PASS: {dtype}")
```

```
PASS: torch.float32
FAIL: torch.int32
PASS: torch.uint8
PASS: torch.bool
```

Yes! We got `torch.bool` working, but curiously `torch.int32` is failing. However that should already honor `low` and `high`? The issue here is that `low` and `high` are given as floating point values, but have to be converted to integer limits internally since `torch.randint` doesn't accepts floats:

```python
>>> torch.randint(-(1 - eps), 1 - eps, (10,), dtype=torch.int32)
TypeError: randint() received an invalid combination of arguments - got (float, float, tuple, dtype=torch.dtype), but expected one of:
[...]
```

This is currently done with

https://github.com/pytorch/pytorch/blob/7a192cc51c7172860efd35413acf9a8b9aafd2c9/torch/testing/_creation.py#L126-L127

Meaning, `-(1 - eps)` is converted to `-1` and `1 - eps` to `1` and thus we get

```pycon
>>> torch.testing.make_tensor(10, dtype=torch.int32, device="cpu", low=-(1 - eps), high=1 - eps)
tensor([ 0,  0,  0,  0, -1,  0, -1,  0, -1,  0], dtype=torch.int32)
```

instead of the all zero tensor that we wanted. Basically, we need to switch to ceiling both `low` `high` to avoid creating values outside of the range explicitly specified by the user. 

But shouldn't we be flooring `high`, i.e. `1 > 1 - eps`? Theoretically yes, but we use `torch.randint(low, high, ...)` afterwards. And since the `high` value is exclusive there, this generates integers in `[low, high)` saving us a `+1`. However, I think for clarity we should accept that.

So why does it work for `torch.uint8` and `torch.bool` above without modification? Because they lucked out in the example I used. Since they are limited to `0` by the dtype, this limits `low` to `0` and thus circumventing the problematic above. For other examples, they would behave exactly as the other integral dtypes.
